src/Formula.elm

```diff
@@ -223,9 +362,29 @@ mapResult f =
     List.foldr (Result.map2 (::) << f) (Ok [])
 
 
+removeQuantifierAndSubstitute : Substitution -> Formula -> Result String Formula
+removeQuantifierAndSubstitute substitution original =
```

Comment:
Nie som si uplne isty, ci toto patri do tohoto modulu, a ci by to nemalo byt az v tvojom tablovaci (robi/mala by robit dost vela specifickych veci, vid nizsie|). Tento modul by mal zverejnovat hlavne `substitute` (well, `substFormula` momentalne) funkciu.

Asi je ale mozne, ze sa najdu aj dalsie vyuzitia, takze asi moze byt aj tu, len treba poriadne definovat co vlastne ma robit, momentalne podla toho co robi je to take divne... podla mna by mala robit nasledovne:
- ak je to `∀ x ....` alebo `∃ x ....`  ak `x` je  jedina premenna v substitucii, tak vrati substituovane [1]
  - ak `x` nie je v subst. (alebo su tam aj ine), tak vrati chybu
- inac vrati chybu (lebo nie je quantifier)

[1] - otazka je, co presne robit ak substitucia obsahuje viac premennych... ak to chces pouzit na validaciu v tablovaci, tak by to malo bud zahodit tie ostatne premenne zo substitucie, alebo vratit chybu, ak tomu das substituciu s viac premennymi (preferujem to druhe).

@ZoltanOnody hodi sa nieco take do tvojej prace? co presne by to malo robit?

Reply:
No takze este raz. Napisem tu viac case-ov. 
Testovat mam `substitute` (po premenovani) alebo tu moju `removeQuantifierAndSubstitute`? Lebo ak mam testovat obe, tak toto vsetko budem musiet definovat pre obe.:

1. Ak moja formula vyzera `∀ x ....` alebo `∃ x ....` a:
    * `x` je v substitucii
        * a je tam viac premennych -> toto sa mi v tableau stat nemoze, ale vo Formule ano a tam to treba testovat. Tu mi napada, ze ta metoda `removeQuantifierAndSubstitute` by mala robit redukciu. Teda nejakym sposobom konverotvat vstup z mojho tableau na vstup pre Formulovsky vseobecny `substitute` a zaroven potom konvertovat jej vystup aby boli pravidla substitucie v tableau dodrzane. Vies o com hovorim?
            Plati to co som nakodila? Ze budem odstranovat kvantifikatory s premennymi kym sa ta premenna nachadza v substitucii?
        Cize ak je tam viac premennych tak v tableau testoch (na `removeQuantifierAndSubstitute`) vyhodim chybu ale vo formule to nevadi. Lenze moze sa mi vobec stat ze z tableau poslem v substitucii viac premennych? Nemoze. Takze ma zmysel to testovat?
        * nie je tam viac premennych -> Ok, substituujem a vratim to.
    * `x` nie je v substitucii -> V tableau je problem a vo formule nie, tak? V testoch pre formulu (`substitute`) to teda zasubstituujem a vratim. V testoch pre moje tableau (`removeQuantifierAndSubstitute`) mam ocakavat chybu. Tak?
2. ak moja formula nema na zaciatku quantifier. -> Toto je v pohode vo vseobecnej Formule (`substitute`) ale nie v tableau. Takze ta metoda `removeQuantifierAndSubstitue` dava celkom aj zmysel. A este mi dava zmysel aj to aby som testy pre `substitute` nechala vo Formule ale tu metodu `removeQuantifierAndSubstitue` presunula do mojho tableau a aj testy k nej napisala.

Co hovoris? @yoyko Dava to zmysel?

Reply:
Nechaj skludom removeQuantifierAndSubstitute v tomto module... myslim, ze sa to hodi aj Zolimu potom...

Co sa tyka testovania:

Vacsina testov by mala byt pre `substitute`, tu chces otestovat vsetky mozne cases so substituciou:
- normalna substitucia,
-  viacero premennych (Ze je to naraz),
- ze viazane premenne to nesubstitije (tj. napriklad x->a do "P(x)&\forall x Q(x)" ma spravit "P(a)&\forall x Q(x)" a pod)
- ze to vrati chybu ked subst. nie je aplikovatelna


`removeQuant...AndSubst...` potom staci otestovat na tie "specialne" pripady, ktora ona riesi:
- vrati chybu ak formula nie je "\forall ? ...." alebo "\exists ? ...",
- vrati chybu ak substitucia nie je prave na tu jednu premennu co je vo vonk. kvant.
- viac menej staci otestovat iba jeden/dva kladne pripady (exists a forall?), hlavne ze korektne urve ten kvantifikator...

Reply:
Oki, dakujem 😃

Reply:
@yoyko , prave som si uvedomia, ze Janciho (@crnkjck ) kod nevracia chybu ked substitucia nie je aplikovatelna. Uz som jancimu aj pisala mail, ale zatial nic.

Napr: `\\forall n \\forall k P(n,k)` a `n->k, k->n` mi nevrati chybu, ale povodnu nesubstituovanu formulu. On si totiz tu:
https://github.com/FMFI-UK-1-AIN-412/elm-formula/pull/2/files#diff-c892343f4203d314590b2465177e1ac6R301 
vsetko nesubstituovatelne (viazane) pomaze zo substitucie, cize tuto:
https://github.com/FMFI-UK-1-AIN-412/elm-formula/pull/2/files#diff-c892343f4203d314590b2465177e1ac6R273 
sa potom nevykona `canSubst`. Lebo v lokalnej premennej `σ` uz nie je to co na zaciatku. a zas keby to odtial nemazem, tak mi zasubstituuje aj viazane...